internal/app/frontend/frontend_service.go

```diff
@@ -45,32 +57,92 @@ func newFrontend(cfg config.View) (*frontendService, error) {
 	return fs, nil
 }
 
-// CreateTicket will create a new ticket, assign a Ticket id to it and put the
-// Ticket in state storage. It will then look through the 'properties' field
-// for the attributes defined as indices the matchmakaking config. If the
-// attributes exist and are valid integers, they will be indexed. Creating a
-// ticket adds the Ticket to the pool of Tickets considered for matchmaking.
+// CreateTicket will create a new ticket, assign an id to it and put it in state
+// storage. It will index the Ticket attributes based on the indexing configuration.
+// Indexing a Ticket adds the it to the pool of Tickets considered for matchmaking.
 func (s *frontendService) CreateTicket(ctx context.Context, req *pb.CreateTicketRequest) (*pb.CreateTicketResponse, error) {
-	return &pb.CreateTicketResponse{}, nil
+	// Generate a ticket id and create a Ticket in state storage
+	ticket := proto.Clone(req.Ticket).(*pb.Ticket)
+	if ticket == nil {
+		logger.Error("invalid argument - ticket cannot be nil")
+		return nil, status.Errorf(codes.InvalidArgument, "ticket cannot be nil")
+	}
+
+	ticket.Id = xid.New().String()
+	err := s.store.CreateTicket(ctx, ticket)
+	if err != nil {
+		logger.WithFields(logrus.Fields{
+			"error":  err.Error(),
+			"ticket": ticket,
+		}).Error("failed to create the ticket")
+		return nil, err
+	}
+
+	err = s.store.IndexTicket(ctx, ticket)
+	if err != nil {
+		logger.WithFields(logrus.Fields{
+			"error":  err.Error(),
+			"ticket": ticket,
+		}).Error("failed to index the ticket")
+		return nil, err
+	}
+
+	return &pb.CreateTicketResponse{Ticket: ticket}, nil
 }
 
-// DeleteTicket removes the Ticket from state storage and from corresponding
-// configured indices. Deleting the ticket stops the ticket from being
-// considered for future matchmaking requests.
+// DeleteTicket removes the Ticket from the configured indexes, thereby removing
+// it from matchmaking pool. It also lazily removes the ticket from state storage.
 func (s *frontendService) DeleteTicket(ctx context.Context, req *pb.DeleteTicketRequest) (*pb.DeleteTicketResponse, error) {
+	// Deindex this Ticket to remove it from matchmaking pool.
+	err := s.store.DeindexTicket(ctx, req.TicketId)
+	if err != nil {
+		logger.WithFields(logrus.Fields{
+			"error": err.Error(),
+			"id":    req.TicketId,
+		}).Error("failed to deindex the ticket")
+		return nil, err
+	}
+
+	// Kick off delete but don't wait for it to complete.
+	go s.deleteTicket(req.TicketId)
 	return &pb.DeleteTicketResponse{}, nil
 }
 
-// GetTicket fetches the ticket associated with the specified Ticket id.
+// deleteTicket is a 'lazy' ticket delete that should be called after a ticket
+// has been deindexed.
+func (s *frontendService) deleteTicket(id string) {
+	err := s.store.DeleteTicket(context.Background(), id)
+	if err != nil {
+		logger.WithFields(logrus.Fields{
+			"error": err.Error(),
+			"id":    id,
+		}).Error("failed to delete the ticket")
+		return
```

Comment:
Do we need a retry logic at here if the call to `store.DeleteTicket` failed?

Reply:
Not here - given that there are no known failure paths we are trying to retry on. At some point, we should explore if there needs to be some garbage collection to catch any data in the database that is stray.